### ISSUE TITLE
Connection not closed (as stated in version 1.0.9 - 2014/10/20)

### ISSUE THREAD

**Issue by quanguyen** [theturtle32/WebSocket-Node#225]
Hi, 

Thanks for the great work on this websocket library. I am using version 1.0.22 (I know this thank to the first lines in **CHANGELOG.md** and **README.md**). After a longtime running, I see that the server still keeps a connection that is supposed to have been dead (because I explicitly powered off the client). Reading my logs, I see that connection never sends _close_ event to my code although the client was turned off.

Reviewing in **CHANGELOG.md**, I learnt that this problem has been fixed since **version 1.0.9**, released on 2014/10/20. Somehow this zombie connection issue is not yet completely fixed? 

Please let me know:
- Am I right to determine my current websocket version by looking at the first lines of **CHANGELOG.md** and **README.md**? I must clarify this important point as I am afraid I am using a version lower than **1.0.9**.
- Has anyone reported this zombie connection since **version 1.0.9**?
- How do I fix this problem?

Thanks,

**Comment by theturtle32** [comment]
> Am I right to determine my current websocket version by looking at the first lines of CHANGELOG.md and README.md? I must clarify this important point as I am afraid I am using a version lower than 1.0.9.

No, you should query the version property:
```javascript
console.log(require('websocket').version)
```

Alternately, look at the version specified in `package.json`.

> Has anyone reported this zombie connection since version 1.0.9?

Not that I'm aware of.

> How do I fix this problem?

I don't know nearly enough about your problem, your code, or your environment to have any hope of answering this question. :-(

If possible please provide a specific test case that can reliably reproduce the behavior that you're seeing and then we can investigating. However, I suspect you are using the library incorrectly.

**Comment by quanguyen** [comment]
Hi Brian,

The version is 1.0.22. I verified by printing out:

> console.log(require('websocket').version)

Below is my code outline:

> "use strict"
> 
> const PORT = SOME_PORT
> 
> var http = require("http")
> var server = http.createServer(function(req, resp) {
>     // Do nothing
> }).listen(PORT, function() {
>     console.log((new Date()) + " Server is listening on port " + PORT)
> })
> 
> var connections = []
> 
> var webSocketServer = require("websocket").server
> var wsServer = new webSocketServer( {
>     httpServer: server
> })
> wsServer.on("request", function(request) {
>     var connection = request.accept(null, request.origin)
> 
>     connection.on("message", function(message) {
>        // Do some logic...
> // If "connection" is not found in "connections", then connections.push(connection) // A
> // else connection.close() // B
>        
>     })
>     
>     connection.on("close", function(code, description) {
>         // find "connection" index in "connections", then connections.splice(index, 1)
>     })
> })

After a longtime running in about a week, when memory reaches over 90MB, one of my client connections keeps disconnecting due to B. I know for sure the code falls into B because before closing connection, I send to the client some message (I did not write it out in the code snippet above for your ease of reading). Obviously, the code falls into B since _connection_ still exists in _connections_ (_close_ event has not been emitted).

Note that, _close_ event works quite well in very first times, but things will happen when it runs in a long time (or maybe memory gets bigger?).

Regards,

**Comment by theturtle32** [comment]
Try pushing your connection in the request handler, not the message handler.

**Comment by quanguyen** [comment]
No, I cannot push my connection in the request handler, which is too early. Before pushing to `connections`, I need to verify it with `// Do some logic...`. If I fail to verify, then I will not push it into the `connections`.

**Comment by valeriakononenko** [comment]
Hi! 
I'm facing the problem, when connection is not removed from websocket.server connections array after connection.close() or connection.drop() were called.

I'm using "websocket": "1.0.35"
Server is created with only  autoAcceptConnections: false
I have only 'request' listener on server

**Comment by theturtle32** [comment]
Can you provide a code example demonstrating the problem?

Brian
